### ISSUE TITLE
torch.nn.CrossEntropyLoss: class weighting changes label_smoothing

### ISSUE THREAD

**Issue by Kait0** [pytorch/pytorch#104776]
### 🐛 Describe the bug

The class weighting in torch.nn.CrossEntropyLoss changes the probabilities that classification networks predict if use together with nonzero label smoothing. This leads to networks learning to predict wrong uncertainties in unbalanced datasets.
In extreme cases / datasets with rare classes the predicted probabilities are so wrong that the wrong class gets predicted.

I wrote a toy example below to reproduce this issue. It is a simple 2 layer network that gets as input a one hot vector with a class label and is supposed to predict the same class. As you can imagine it gets 100 % training accuracy with standard setting and predicts uncertainties according to the correct label smoothed probabilities.
The code snippet contains 4 tests with different class weighting.

When upweighting one of the classes (together with label smoothing) the training accuracy drops to 0.0008 because the network always puts 0.9147% probability on the rare class. This is because the label smoothing on the incorrect class is increased by the class weighting.

Training without label smoothing but with class weighting again gives 100% accuracy.

My expectation would be that class weighting does not change how much probability the label smoothing puts on the wrong classes.

Example:
(I use sklearn to automatically compute balanced class weights here but one can of course also set them manually.)

```Python
import torch
from torch import nn
from torch import optim
import torch.nn.functional as F
import numpy as np
from sklearn.utils.class_weight import compute_class_weight


num_classes = 4

# Test 1: balanced dataset + label smoothing -> balance class weights -> no problem
#probabilities = [0.25, 0.25, 0.25, 0.25]
#label_smoothing = 0.1

# Test 2: imbalanced dataset + label smoothing -> one class has 8x higher weight -> wrong uncertainties
#probabilities = [0.32, 0.04, 0.32, 0.32]
#label_smoothing = 0.1

# Test 3: rare class + label smoothing -> one class 333x higher weight -> wrong uncertainties + wrong predictions
probabilities = [0.333, 0.001, 0.333, 0.333]
label_smoothing = 0.1

# Test 4: rare class no label smoothing -> one class ~300x higher weight -> no problem
#probabilities = [0.333, 0.001, 0.333, 0.333]
#label_smoothing = 0.0

model = torch.nn.Sequential(nn.Linear(num_classes, 32),
                            nn.ReLU(inplace=True),
                            nn.Linear(32, num_classes))
optimizer = optim.AdamW(model.parameters(), lr=0.0003)

dataset_size = 10000
options = np.arange(num_classes)
# Normalize to make sure we have valid probabilities
probabilities = [float(i)/sum(probabilities) for i in probabilities]
print('Class probabilities in dataset', probabilities)
inputs = []
labels = []
labels_numpy = []

for _ in range(dataset_size):
  sample = np.random.choice(options, p=probabilities)
  labels_numpy.append(sample)
  tensor_sample = torch.from_numpy(np.array(sample))
  labels.append(tensor_sample)
  inputs.append(F.one_hot(tensor_sample, num_classes).to(torch.float32))



inputs = torch.stack(inputs)
labels = torch.stack(labels)
model.train()
class_weights = torch.tensor(compute_class_weight(class_weight='balanced', classes=np.arange(4), y=labels_numpy),
                             dtype=torch.float32)
print('Class weights', class_weights)
ce_loss = nn.CrossEntropyLoss(weight=class_weights, label_smoothing=label_smoothing)
optimizer.zero_grad()

for epoch in range(5000):
  prediction = model(inputs)
  loss = ce_loss(prediction, labels)
  if epoch % 100 == 0:
    print('Epoch:', epoch, loss)
  loss.backward()
  optimizer.step()
  optimizer.zero_grad()


final_logits = model(inputs)
final_prob = torch.nn.functional.softmax(final_logits, dim=1)
final_pred = torch.argmax(final_prob, dim=1)
print(final_prob[:10])
print(final_pred[:10])
print(labels[:10])
test = labels == final_pred
print('Accuracy:', sum(test) / len(test))
```

### Versions

[pip3] numpy==1.21.6
[pip3] pytorch-lightning==0.7.1
[pip3] torch==1.12.1+cu113
[pip3] torchaudio==0.12.1+cu113
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.13.1+cu113
[conda] numpy                     1.21.6                   pypi_0    pypi
[conda] pytorch-lightning         0.7.1                    pypi_0    pypi
[conda] torch                     1.12.1+cu113             pypi_0    pypi
[conda] torchaudio                0.12.1+cu113             pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchvision               0.13.1+cu113             pypi_0    pypi

**Comment by nathanfranke** [comment]
I can reproduce this issue with CrossEntropyLoss.

```py
for multiplier, weight in [(10.0, torch.tensor([0.1, 0.9])), (20.0, torch.tensor([0.01, 0.99])), (1.0, torch.tensor([1.0, 1.0])), (1.0, torch.tensor([1.0, 10.0])), (0.2, torch.tensor([1.0, 100.0]))]:
    print(f"Weight {weight[0].item():.2f} {weight[1].item():.2f}", end="")
    target = torch.tensor([1.0, 0.0])
    for i in range(21):
        outputs = torch.tensor([i * 0.05, 1.0 - i * 0.05])
        criterion = torch.nn.CrossEntropyLoss(weight=weight, label_smoothing=0.1)
        cost = criterion(outputs.unsqueeze(0), target.unsqueeze(0))
        display = cost.item() * multiplier
        print(f", {display:.5f}", end="")
    print()
```

Output:
```
Weight 0.10 0.90, 1.38857, 1.33262, 1.27954, 1.22946, 1.18248, 1.13871, 1.09822, 1.06110, 1.02739, 0.99716, 0.97041, 0.94716, 0.92739, 0.91110, 0.89822, 0.88871, 0.88248, 0.87946, 0.87954, 0.88262, 0.88857
Weight 0.01 0.99, 0.55965, 0.57356, 0.58990, 0.60876, 0.63024, 0.65441, 0.68136, 0.71114, 0.74380, 0.77939, 0.81791, 0.85939, 0.90380, 0.95114, 1.00136, 1.05441, 1.11024, 1.16876, 1.22990, 1.29356, 1.35965
Weight 1.00 1.00, 1.26326, 1.19615, 1.13110, 1.06819, 1.00749, 0.94908, 0.89302, 0.83936, 0.78814, 0.73940, 0.69315, 0.64940, 0.60814, 0.56936, 0.53302, 0.49908, 0.46749, 0.43819, 0.41110, 0.38615, 0.36326
Weight 1.00 10.00, 1.40423, 1.34967, 1.29810, 1.24962, 1.20436, 1.16241, 1.12387, 1.08882, 1.05730, 1.02938, 1.00506, 0.98438, 0.96730, 0.95382, 0.94387, 0.93741, 0.93436, 0.93462, 0.93810, 0.94467, 0.95423
Weight 1.00 100.00, 0.56278, 0.57697, 0.59361, 0.61279, 0.63461, 0.65915, 0.68649, 0.71668, 0.74979, 0.78583, 0.82485, 0.86683, 0.91179, 0.95968, 1.01049, 1.06415, 1.12061, 1.17979, 1.24161, 1.30597, 1.37278
```

![image](https://github.com/pytorch/pytorch/assets/14253836/154ae683-5498-46df-bb15-fbf86da054ff)

**Comment by jbschlosser** [comment]
I haven't had time to look into this in too much detail, but there's a specific perf optimization we're doing for the label smoothing calculation that I'm sure implicitly requires unit weights to be correct. I'll see if I can get time soon to identify a fix.
